File: l33tcode/word-search.py
class Solution:
    def exist(self, board, word):
        if not board or not word:
            return False

        visited = [[False for _ in board[0]] for _ in board]
        def find(row, col, pos):
            if pos == len(word) - 1:
                return True

            visited[row][col] = True
            result = False

            for new_row, new_col in [
                (row - 1, col),
                (row, col - 1),
                (row + 1, col),
                (row, col + 1),
            ]:
                if 0 <= new_row < len(board) and \
                   0 <= new_col < len(board[0]) and \
                   board[new_row][new_col] == word[pos + 1] and \
                   not visited[new_row][new_col] and \
                   find(new_row, new_col, pos + 1):
                    result = True
                    break

            visited[row][col] = False
            return result

        for row in range(len(board)):
            for col in range(len(board[0])):
                if board[row][col] == word[0] and find(row, col, 0):
                    return True

        return False

File: l33tcode/word-search.go
package main

func exist(board [][]byte, word string) bool {
	rows, cols := len(board), len(board[0])
	createVisited := func(rows, cols int) [][]bool {
		result := [][]bool{}

		for row := range make([]bool, rows) {
			result = append(result, []bool{})
			for _ = range make([]bool, cols) {
				result[row] = append(result[row], false)
			}
		}
		return result
	}
	neighbours := func(row, col int, visited *[][]bool) [][]int {
		tmp := [][]int{}
		tmp = append(tmp, []int{row - 1, col})
		tmp = append(tmp, []int{row + 1, col})
		tmp = append(tmp, []int{row, col - 1})
		tmp = append(tmp, []int{row, col + 1})
		result := [][]int{}
		for _, neighbour := range tmp {
			neighRow, neighCol := neighbour[0], neighbour[1]
			if neighRow >= 0 && neighRow < rows && neighCol >= 0 && neighCol < cols {
				if !(*visited)[neighRow][neighCol] {
					result = append(result, []int{neighRow, neighCol})
				}
			}
		}
		return result
	}
	var backtrack func(row, col int, wordPos int, visited *[][]bool) bool
	backtrack = func(row, col int, wordPos int, visited *[][]bool) bool {
		if word[wordPos] != board[row][col] {
			return false
		}

		if wordPos == len(word)-1 {
			return true
		}

		for _, neighbour := range neighbours(row, col, visited) {
			neighRow, neighCol := neighbour[0], neighbour[1]
			(*visited)[row][col] = true
			if backtrack(neighRow, neighCol, wordPos+1, visited) {
				return true
			}
			(*visited)[row][col] = false
		}
		return false
	}

	visited := createVisited(rows, cols)
	for row := range board {
		for col := range board[0] {
			if backtrack(row, col, 0, &visited) {
				return true
			}
		}
	}
	return false
}
